x-pack/plugins/stack_connectors/server/connector_types/crowdstrike/crowdstrike.ts

```diff
@@ -0,0 +1,179 @@
+  private registerSubActions() {
+    this.registerSubAction({
+      name: SUB_ACTION.GET_AGENT_DETAILS,
+      method: 'getAgentDetails',
+      schema: CrowdstrikeGetAgentsParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.HOST_ACTIONS,
+      method: 'executeHostActions',
+      schema: CrowdstrikeHostActionsParamsSchema,
+    });
+  }
+
+  public async executeHostActions({ alertIds, ...payload }: CrowdstrikeHostActionsParams) {
+    return this.crowdstrikeApiRequest({
+      url: this.urls.hostAction,
+      method: 'post',
+      params: {
+        action_name: payload.command,
+      },
+      data: {
+        ids: payload.ids,
+        ...(payload.actionParameters
+          ? {
+              action_parameters: Object.entries(payload.actionParameters).map(([name, value]) => ({
+                name,
+                value,
+              })),
+            }
+          : {}),
+      },
+      paramsSerializer,
+      responseSchema: CrowdstrikeHostActionsResponseSchema,
+    });
+  }
+
+  public async getAgentDetails(
+    payload: CrowdstrikeGetAgentsParams
+  ): Promise<CrowdstrikeGetAgentsResponse> {
+    return this.crowdstrikeApiRequest({
+      url: this.urls.agents,
+      method: 'GET',
+      params: {
+        ids: payload.ids,
+      },
+      paramsSerializer,
+      responseSchema: CrowdstrikeGetAgentsResponseSchema,
+    });
+  }
+
+  private async getTokenRequest() {
+    const response = await this.request<CrowdstrikeGetTokenResponse>({
+      url: this.urls.getToken,
+      method: 'post',
+      headers: {
+        accept: 'application/json',
+        'Content-Type': 'application/x-www-form-urlencoded',
+        authorization: 'Basic ' + this.base64encodedData,
+      },
+      responseSchema: CrowdstrikeGetTokenResponseSchema,
+    });
+    const token = response.data?.access_token;
+    if (token) {
+      // Clear any existing timeout
+      clearTimeout(CrowdstrikeConnector.tokenExpiryTimeout);
+
+      // Set a timeout to reset the token after 29 minutes (it expires after 30 minutes)
+      CrowdstrikeConnector.tokenExpiryTimeout = setTimeout(() => {
+        CrowdstrikeConnector.token = null;
+      }, 29 * 60 * 1000);
+    }
+    return token;
+  }
+
+  private async crowdstrikeApiRequest<R extends CrowdstrikeBaseApiResponse>(
+    req: SubActionRequestParams<R>,
+    retried?: boolean
+  ): Promise<R> {
+    try {
+      if (!CrowdstrikeConnector.token) {
+        CrowdstrikeConnector.token = (await this.getTokenRequest()) as string;
+      }
+
+      const response = await this.request<R>({
+        ...req,
+        headers: {
+          ...req.headers,
+          Authorization: `Bearer ${CrowdstrikeConnector.token}`,
+        },
+      });
+
+      return response.data;
+    } catch (error) {
+      if (error.code === 401 && !retried) {
+        CrowdstrikeConnector.token = null;
+        return this.crowdstrikeApiRequest(req, true);
+      }
+
+      throw new Error(error.message);
```

Comment:
can we improve the Error response here (I been meaning to do the same for SentinelOne).
One thing that I have found to be very difficult to work with is the fact that the Response body of the failure is not accessible and that usually provide more information than just (for example): `Error: Status code: 400. Message: API Error: Bad Request`

IMO, it would be nice to have a Crowdstrike specific Error class (sub-class from `Error`) that could include additional data associated with the HTTP failure. I'm not suggesting the entire `Response` object be returned, as that would expose some data that we definitely  don't want to expose (API keys).

Reply:
something like this #8e816f0d65fe6a99011bbd08e5ac20f3574e6c13 ?